### PR TITLE
Request/csverror typescript support

### PR DESCRIPTION
**Pull request by drl-max**
I've added a Typescript definition for `CsvError` to address issue #295.

I also added the `CsvErrorCode` type to enumerate the possible error messages. Definitely not required, just thought it would be nice. I can remove it if that's preferred. Also, let me know if there are other tests you'd like to see performed.

### CODE REVIEW

**Comment by wdavidw** [comment]
Isn't `enum` a better fit than `type` for those error codes ?

**Comment by drl-max** [comment]
Full disclosure, I am _not_ an expert at Typescript, so please take the following assessment with a grain-of-salt.

**tl;dr** - Going with the `type` approach means only the Typescript typings will be changed. Going with `enum` will require the module itself to export the same data (as well as having the Typescript definition).

I've, personally, had issues with trying to use `enum` in the past but, since you asked, I tried it out.  Sadly, it didn't seem to be working in the unit tests and I couldn't figure out why until I realized that these `index.d.ts` files do nothing more than _describe_ the underlying module. Defining the codes as a `type`, will only apply to other Typescript users (limits possible values to only the ones listed) but defining an `enum` requires that there is an equivalently structured property being exported from the module. 

For kicks, I'll put up another PR for the `enum` approach.

**Comment by wdavidw** [comment]
Does any one with Typescript expertise and passing through this PR have any input to share with us ?

**Comment by wdavidw** [comment]
My preference will go for the enum variation. I really wish someone would confirm/validate this however. Maybe @yonilerner ?

**Comment by yonilerner** [comment]
I think the union string literal type approach (ie. this PR, not #297) is preferred, unless you actually *want* to introduce an enum into the JS itself, but presumably there was a reason for using string literals in the JS? Regardless, just here to say that this approach using a `type` representing a union of string literals is quite idiomatic for TS

**Comment by wdavidw** [comment]
I don't mind introducing an array in JS but if you confirm that using type is better, then lets go for it.

**Comment by drl-max** [comment]
@wdavidw : Would you still like those changes you mentioned in #297 applied to this PR?

**Comment by wdavidw** [comment]
No, following your first proposal and @yonilerner comment, Let's go for this one.

**Comment by wdavidw** [comment]
If you feel like doing more, `lib/sync` shall also honor the errors.

**Comment by wdavidw** [comment]
And thank you for your contribution. Version 4.12.0 is published.
